radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -20,55 +23,289 @@ mod test;
+            (
+                root.hash,
+                Payload::ReNodeModule(ReNodeModulePayload {
+                    re_node_id: re_node_module.re_node_id,
+                    node_mode_id: re_node_module.node_module_id,
+                    substates_root: root.node,
+                }),
+            )
+        }),
+    }
+}
+
+fn to_substate_change(change: IdChange<SubstateOffset, Hash>) -> LeafChange {
+    LeafChange {
+        key_hash: hash(scrypto_encode(&change.id).unwrap()),
+        new_payload: change
+            .changed
+            .map(|value_hash| (value_hash, Payload::Substate(change.id))),
+    }
+}
+
+fn break_tuple<A, B>(tuple: Option<(A, B)>) -> (Option<A>, Option<B>) {
+    match tuple {
+        Some((a, b)) => (Some(a), Some(b)),
+        None => (None, None),
+    }
+}
+
+struct NestedTreeStore<'s, S> {
+    underlying: &'s mut S,
+    parent_path: NibblePath,
+    current_root: Option<TreeNode<Payload>>,
+    new_root: Option<TreeNode<Payload>>,
+}
+
+impl<'s, S> NestedTreeStore<'s, S> {
+    pub fn new(
+        underlying: &'s mut S,
+        re_node_module: &ReNodeModule,
+        root: Option<TreeNode<SubstateOffset>>,
+    ) -> NestedTreeStore<'s, S> {
+        NestedTreeStore {
+            underlying,
+            parent_path: NibblePath::new_even(
+                hash(scrypto_encode(re_node_module).unwrap()).to_vec(),
+            ),
+            current_root: root.map(|tree_node| {
+                tree_node.map_payload(|substate_offset| Payload::Substate(substate_offset))
+            }),
+            new_root: None,
+        }
+    }
+
+    pub fn extract_new_root(&mut self) -> TreeNode<Payload> {
+        self.new_root
+            .take()
+            .expect("no new root stored into the nested tree")
+    }
+
+    fn prefixed(&self, key: &NodeKey) -> NodeKey {
+        NodeKey::new(
+            key.version(),
+            NibblePath::from_iter(
+                self.parent_path
+                    .nibbles()
+                    .chain(key.nibble_path().nibbles()),
+            ),
+        )
+    }
+}
+
+impl<'s, S: ReadableTreeStore> ReadableTreeStore for NestedTreeStore<'s, S> {
+    fn get_node(&self, key: &NodeKey) -> Option<TreeNode<Payload>> {
+        if key.nibble_path().is_empty() {
+            self.current_root.clone()
+        } else {
+            self.underlying.get_node(&self.prefixed(key))
+        }
+    }
+}
+
+impl<'s, S: WriteableTreeStore> WriteableTreeStore for NestedTreeStore<'s, S> {
+    fn insert_node(&mut self, key: NodeKey, node: TreeNode<Payload>) {
+        if key.nibble_path().is_empty() {
+            self.new_root = Some(node);
+        } else {
+            self.underlying.insert_node(self.prefixed(&key), node);
+        }
+    }
+
+    fn record_stale_node(&mut self, key: NodeKey) {
+        if key.nibble_path().is_empty() {
+            self.current_root = None;
+        } else {
+            self.underlying.record_stale_node(self.prefixed(&key));
```

Comment:
Hmmm this is something I didn't understand initially -- I assumed we'd have both a leaf node in the parent tree, and a child root node - but it sounds like maybe the leaf node in the parent tree is a wrapper around the root of the child tree? In which case, this is awesome as it's more efficient from a node storage perspective :D.

But perhaps it's something worth adding to the Rust docs more explicitly?

Reply:
> but it sounds like maybe the leaf node in the parent tree is a wrapper around the root of the child tree? In which case, this is awesome as it's more efficient from a node storage perspective :D.

Yes, it's exactly like that.
It saves ~0.1% of storage _(number not actually measured)_ and also saves 1 hop during tree traversal _(number actually exact)_.

But I did it actually to avoid a possible awkward DB key clash:
Imagine we have a ReNode leaf of hash(ReNodeId+ModuleId) = `0xa1b2c3`.
And unfortunately a sibling hash also exists (let's say `0xa1b2c4`).
It means the ReNode leaf will have to use a all nibbles of the hash in its DB key, like `key=v7|a1b2c3`.
And guess what is the DB key of a root node of the substates' subtree of that ReNode? yup, it is `key=v7|a1b2c3` (i.e. an empty nibble path [as for every root], prefixed by its ReNode's full nibble path). 
This has very low probability, but it hints that conceptually the ReNode leaf has the same identity as the Substate's subtree root. So I adopted that concept.

_(of course the unlikely key clash could be battled e.g. by adding some separator nibble, or a ReNode/Substate discriminator nibble at the beginning, but these are solutions from the "ugly sentinel" territory)._

Reply:
I added Rust doc (on `ReNodeModulePayload`) for it, PTAL.

Reply:
It totally makes sense that they're the same thing, I really like it tbh :D